language/diem-vm/src/diem_transaction_executor.rs

```diff
@@ -502,33 +502,46 @@ impl DiemVM {
         {
             return Ok(discard_error_vm_status(e));
         };
-
-        let change_set = match txn.payload() {
-            TransactionPayload::WriteSet(writeset_payload) => {
-                match self.execute_writeset(
-                    remote_cache,
-                    writeset_payload,
-                    Some(txn_data.sender()),
-                    log_context,
-                ) {
-                    Ok(change_set) => change_set,
-                    Err(e) => return e,
+        self.execute_writeset_transaction(
+            remote_cache,
+            match txn.payload() {
+                TransactionPayload::WriteSet(writeset_payload) => writeset_payload,
+                TransactionPayload::Module(_) | TransactionPayload::Script(_) => {
+                    log_context.alert();
+                    error!(*log_context, "[diem_vm] UNREACHABLE");
+                    return Ok(discard_error_vm_status(VMStatus::Error(
+                        StatusCode::UNREACHABLE,
+                    )));
                 }
-            }
-            TransactionPayload::Module(_) | TransactionPayload::Script(_) => {
-                log_context.alert();
-                error!(*log_context, "[diem_vm] UNREACHABLE");
-                return Ok(discard_error_vm_status(VMStatus::Error(
-                    StatusCode::UNREACHABLE,
-                )));
-            }
+            },
+            TransactionMetadata::new(&txn),
+            log_context,
+        )
+    }
+
+    pub fn execute_writeset_transaction(
+        &mut self,
+        remote_cache: &StateViewCache<'_>,
+        writeset_payload: &WriteSetPayload,
+        txn_data: TransactionMetadata,
+        log_context: &impl LogContext,
+    ) -> Result<(VMStatus, TransactionOutput), VMStatus> {
+        let change_set = match self.execute_writeset(
+            remote_cache,
+            writeset_payload,
+            Some(txn_data.sender()),
+            log_context,
+        ) {
+            Ok(change_set) => change_set,
+            Err(e) => return e,
         };
 
-        // Emit the reconfiguration event
+        // Run the epilogue function.
+        let mut session = self.0.new_session(remote_cache);
```

Comment:
...and related to the previous comment, I believe the session allocated at the beginning of process_writeset_transaction() will be unused and you can remove that. (It was removed in #6891 but that isn't included on the branch.)

Reply:
The session is used for epilogue so we probably cant remove that.

Reply:
I guess you mean "prologue" in that comment. I see your point. I think this is OK, but just to point out, there is a minor change in behavior here because we'll be using different sessions to run the prologue and epilogue. We had that on master between #6891 and #7175 and it seemed fine except for the performance regression, which shouldn't be a concern for WriteSet transactions.